### PR TITLE
Fix RedBlackTree impl/test #81

### PR DESCRIPTION
**Pull request by PatrykOlejniczak**
Hi,
Referring to [issue](https://github.com/aalhour/C-Sharp-Algorithms/issues/81)

[Tests](https://github.com/aalhour/C-Sharp-Algorithms/blob/master/UnitTest/DataStructuresTests/RedBlackTreeTest.cs) related with [RedBlackTree](https://github.com/aalhour/C-Sharp-Algorithms/blob/master/DataStructures/Trees/RedBlackTree.cs) did not pass so I decided to take a closer look at them. 

The method of inserting new elements had an error in setting the color while repairing the properties of red black. It was fixed in one line [link](https://github.com/aalhour/C-Sharp-Algorithms/compare/master...PatrykOlejniczak:master#diff-c17fa9ff8bceae134ebbd3fd98e06209L236)

When adding new tests, there were also problems with removing nodes. 😢 
I also decided to fix it, but despite the sincere intentions, I had to rewrite the entire [_remove](https://github.com/aalhour/C-Sharp-Algorithms/compare/master...PatrykOlejniczak:master#diff-c17fa9ff8bceae134ebbd3fd98e06209R461) method, but `_adjustTreeAfterRemoval` remained unchanged and it works fine. 
I wanted to avoid previous convention naming the variables `node1, node2` so that the method would be more readable. Unfortunately I did not find enough short and legible names for [X](https://github.com/aalhour/C-Sharp-Algorithms/compare/master...PatrykOlejniczak:master#diff-c17fa9ff8bceae134ebbd3fd98e06209R468) and [Y](https://github.com/aalhour/C-Sharp-Algorithms/compare/master...PatrykOlejniczak:master#diff-c17fa9ff8bceae134ebbd3fd98e06209R483), so I added a comment about what it means, I'm open to change.

I also added properties in base BST [HasOnlyRightChild ](https://github.com/aalhour/C-Sharp-Algorithms/compare/master...PatrykOlejniczak:master#diff-77767a2584b7b66d9c4085b71681e38eR66) and [HasOnlyLeftChild ](https://github.com/aalhour/C-Sharp-Algorithms/compare/master...PatrykOlejniczak:master#diff-77767a2584b7b66d9c4085b71681e38eR79) that allow you to write shorten `if` conditions.

During the implementation I was supported by the Cormen pseudocode, so i hope it will be essentially correct. If you have any ideas for more tests, I can add them.

I also added todos [1](https://github.com/aalhour/C-Sharp-Algorithms/compare/master...PatrykOlejniczak:master#diff-e4cbe43925b19dfec6df7243e05dd5acR211) [2](https://github.com/aalhour/C-Sharp-Algorithms/compare/master...PatrykOlejniczak:master#diff-e4cbe43925b19dfec6df7243e05dd5acR374) for the future, I did not want to introduce these changes within this pull-request.

If there are any comments I will try to improve it. Sorry for my english if something is not clear please ask.

Regards,

### CODE REVIEW

**Comment by sj66phone** [comment]
Merge remote-tracking branch 'upstream/development'

**Comment by aalhour** [comment]
Hi @PatrykOlejniczak,

I finally got some time to take a deeper look at this. I think the PR is quite substantial and I have two questions:
  1. Can you please point me to the Cormen pseudocode you used? My reference was the [Sedgewick version](https://algs4.cs.princeton.edu/home/).
  2. Can you tell me why this test case doesn't rebalance around #7? [Click here](https://github.com/aalhour/C-Sharp-Algorithms/pull/82/files#diff-e4cbe43925b19dfec6df7243e05dd5acR332-R340).

File: DataStructures/Trees/RedBlackTree.cs [hunk 1]
```diff
@@ -455,80 +454,104 @@ protected override bool _remove(BSTNode<TKey> nodeToDelete)
...
-                node1 = (RedBlackTreeNode<TKey>)base._findNextLarger(nodeToDelete);
+                return false;
             }
 
-            // Left child case
-            if (node1.HasLeftChild)
+            if (!nodeToDelete.HasChildren)
             {
-                node2 = node1.LeftChild;
+                Root = null;
```
Comment (aalhour) [inline DataStructures/Trees/RedBlackTree.cs]: Why do you nullify the Root of the tree if the node we want to delete doesn't have any children? Can you please add a comment explaining the reason behind this?

File: DataStructures/Trees/RedBlackTree.cs [hunk 2]
```diff
@@ -455,80 +454,104 @@ protected override bool _remove(BSTNode<TKey> nodeToDelete)
...
             }
             else
             {
-                node2 = node1.RightChild;
-            }
+                // X it's node that will become move to original nodeToDelete position in the tree.
```
Comment (aalhour) [inline DataStructures/Trees/RedBlackTree.cs]: Maybe re-write as:
> X is the node we will replace with the `nodeToDelete` in the tree once we remove it.

File: DataStructures/Trees/RedBlackTree.cs [hunk 3]
```diff
@@ -455,80 +454,104 @@ protected override bool _remove(BSTNode<TKey> nodeToDelete)
...
                 }
-            }
-            else
-            {
-                Root = node2;
-                Root.Parent = null;
-            }
+                else
+                {
+                    // Y it's node that will become move to original X position in the tree.
```
Comment (aalhour) [inline DataStructures/Trees/RedBlackTree.cs]: Maybe re-write as:
> Y is the node we will replace with the `X` in the tree once we move it to the `nodeToDelete` position.

File: DataStructures/Trees/RedBlackTree.cs (hunk 1 above)
Comment (PatrykOlejniczak) [reply DataStructures/Trees/RedBlackTree.cs]: Yes i will fix that, i don't remember why i done that...

File: DataStructures/Trees/RedBlackTree.cs (hunk 2 above)
Comment (PatrykOlejniczak) [reply DataStructures/Trees/RedBlackTree.cs]: Fixed.

File: DataStructures/Trees/RedBlackTree.cs (hunk 3 above)
Comment (PatrykOlejniczak) [reply DataStructures/Trees/RedBlackTree.cs]: Fixed.

File: DataStructures/Trees/RedBlackTree.cs (hunk 1 above)
Comment (PatrykOlejniczak) [reply DataStructures/Trees/RedBlackTree.cs]: Fixed.
I added condition to check that element is root before assign, and created new unit test that cover it.
In line 480 is covered case when we remove node that haven't children and is't root. 

Thanks for your attention, I don't know how I missed such an obvious case.

**Comment by PatrykOlejniczak** [comment]
Hi @aalhour 
Fixed your suggestions thank for attention and time spend to review that.

Back to your questions:
1. I refer/use Cormen Introduce to Algorithms [link](https://mitpress.mit.edu/books/introduction-algorithms-third-edition) but i didn't rewrite 1:1, added more test and then do some changes to cover that.
2. My bad again, it should be rebalanced, sorry I'll fix that.

**Comment by aalhour** [comment]
@karv - would you mind taking a look at this one too?

**Pull request merged**